client/src/main/java/org/eclipse/hono/client/impl/AbstractHonoClient.java

```diff
@@ -165,8 +165,17 @@ protected static final void setApplicationProperties(final Message msg, final Ma
             sender.setQoS(qos);
             sender.openHandler(senderOpen -> {
                 if (senderOpen.succeeded()) {
-                    LOG.debug("sender open [{}]", sender.getRemoteTarget());
-                    result.complete(senderOpen.result());
+                    LOG.debug("sender open [{}] sendQueueFull [{}]", sender.getRemoteTarget(), sender.sendQueueFull());
+                    // wait on credits a little time, if not already given TODO: configurable waiting time
+                    if(sender.sendQueueFull()) {
```

Comment:
I think this is exactly the right place to put the wait cycle handling :-) Great idea to first check if we already have credit before doing the wait!
There is a whitespace missing before opening bracket, please adjust your editor settings accordingly